### PR TITLE
Task, ClusterTask and Pipeline auto-conversion from v1alpha1 to v1alpha2 🎋

### PR DESCRIPTION
**Pull request by vdemeester**
# Changes

This adds:
- auto-conversion methods and tests for Task types in v1alpha1
and v1alpha2. In order to reduce the number of duplicated fields, we
are using v1alpha2.TaskSpec embedded in v1alpha1.TaskSpec so that both
types are compatible.
- auto-conversion methods and tests for ClusterTask
- auto-conversion methods and tests for Pipeline

/cc @sbwsg @bobcatfish @afrittoli 

Signed-off-by: Vincent Demeester <vdemeest@redhat.com>

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a new binary/image to the `cmd` dir](../cmd), please update
  [the release Task](../tekton/publish.yaml) to build and release this image.

## Reviewer Notes

If [API changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md) are included, [additive changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#additive-changes) must be approved by at least two [OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS) and [backwards incompatible changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-incompatible-changes) must be approved by [more than 50% of the OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS), and they must first be added [in a backwards compatible way](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-compatible-changes-first).

### CODE REVIEW

File: pkg/apis/pipeline/v1alpha1/pipeline_conversion.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+)
+
+var _ apis.Convertible = (*Pipeline)(nil)
+
+// ConvertUp implements api.Convertible
+func (source *Pipeline) ConvertUp(ctx context.Context, obj apis.Convertible) error {
+	switch sink := obj.(type) {
+	case *v1alpha2.Pipeline:
+		sink.ObjectMeta = source.ObjectMeta
+		return nil // source.Spec.ConvertUp(ctx, &sink.Spec)
```
Comment (vdemeester) [inline pkg/apis/pipeline/v1alpha1/pipeline_conversion.go]: This is a placeholder :angel: I needed to define at least 2 Convertible types to some tests :stuck_out_tongue_closed_eyes:

File: pkg/apis/pipeline/v1alpha1/task_types.go [hunk 2]
```diff
@@ -46,6 +45,8 @@ func (t *Task) Copy() TaskInterface {
 
 // TaskSpec defines the desired state of Task.
 type TaskSpec struct {
+	v1alpha2.TaskSpec `json:",inline"`
```
Comment (vdemeester) [inline pkg/apis/pipeline/v1alpha1/task_types.go]: This is not required per-se, it just simplify the conversion using embedded struct *and* it doesn't change on the api side (except user of the go library)

Comment (vdemeester) [reply pkg/apis/pipeline/v1alpha1/task_types.go]: If you look at `Pipeline`, it is not using embedding as we can't really do that (same `task` field targeting different type)

File: pkg/apis/pipeline/v1alpha1/task_conversion.go [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+			for i, param := range source.Inputs.Params {
+				sink.Params[i] = *param.DeepCopy()
+			}
+		}
+		if len(source.Inputs.Resources) > 0 {
+			if sink.Resources == nil {
+				sink.Resources = &v1alpha2.TaskResources{}
+			}
+			if len(source.Inputs.Resources) > 0 && source.Resources != nil && len(source.Resources.Inputs) > 0 {
+				// This shouldn't happen as it shouldn't pass validation but just in case
```
Comment (sbwsg) [inline pkg/apis/pipeline/v1alpha1/task_conversion.go]: :+1:

File: pkg/apis/pipeline/v1alpha1/task_conversion.go [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+				return apis.ErrMultipleOneOf("inputs.resources", "resources.inputs")
+			}
+			sink.Resources.Inputs = make([]v1alpha2.TaskResource, len(source.Inputs.Resources))
```
Comment (sbwsg) [inline pkg/apis/pipeline/v1alpha1/task_conversion.go]: :tada:

File: pkg/apis/pipeline/v1alpha1/task_types.go (hunk 2 above)
Comment (sbwsg) [reply pkg/apis/pipeline/v1alpha1/task_types.go]: aha, so this is how TaskSpec.ConvertDown is able [to copy into the v1alpha1.Params and Resources fields](https://github.com/tektoncd/pipeline/pull/2002/files#diff-35489a0872e9373a16f7115dd2618c70R120)?

:+1:

**Comment by sbwsg** [comment]
I don't totally understand the need for v1alpha1 ConvertDown funcs.  What is their use?

**Comment by vdemeester** [comment]
> I don't totally understand the need for v1alpha1 ConvertDown funcs. What is their use?

I did write about this in the initial pull request #1529 : 
- Only one version of the CRD can be stored, so `v1alpha1` is the one
stored
- All the version of the CRD **must** be schema compatible (before 1.16)
  - `v1alpha1` gets **all** the fields, meaning :
    - the new ones (from `v1alpha2`, …)
    - the old deprecated ones (that will be removed from `v1alpha2`)
  - we may *only* document the "frozen" `v1alpha1` field (and not the
    new one that are coming from the other api versions)

The main reason to `ConvertDown` is because we do store it in `v1alpha1`. So if a user send a `v1alpha1` struct, it will convert it up and down, to be `v1alpha2` but still stored as a `v1alpha1` struct (kinda).

**Comment by sbwsg** [comment]
> I did write about this in the initial pull request #1529 :

:angel: 

> The main reason to ConvertDown is because we do store it in v1alpha1. So if a user send a v1alpha1 struct, it will convert it up and down, to be v1alpha2 but still stored as a v1alpha1 struct (kinda).

🤯 Brilliant.  Thanks!

**Comment by chmouel** [comment]
the golang ci test is fallaciously green since it is actually failing but exit 0 

```
0206 09:56:42.536] ===============================
I0206 09:56:42.536] ==== RUNNING GOLANGCI-LINT ====
I0206 09:56:42.536] ===============================
I0206 09:57:44.392] level=warning msg="[runner] Can't run linter goanalysis_metalinter: ctrlflow: failed prerequisites: inspect@github.com/tektoncd/pipeline/pkg/artifacts [github.com/tektoncd/pipeline/pkg/apis/pipeline/v1alpha1.test]"
```
This seems a recurring errors in golang-ci which i have updated the bug with our issue in here : 

https://github.com/golangci/golangci-lint/issues/827#issuecomment-582916152

**Comment by ernado** [comment]
Hey, can you please check what actual golangci version are you using?
I suspect that is something like `v1.21`, but there is already `v1.23.3` with important fixes.

**Comment by vdemeester** [comment]
> Hey, can you please check what actual golangci version are you using?
> I suspect that is something like `v1.21`, but there is already `v1.23.3` with important fixes.

See https://github.com/tektoncd/plumbing/pull/216 :angel:

**Pull request merged**